frontend/controller/app/identity.js

```diff
@@ -0,0 +1,252 @@
+      throw new GIErrorUIRuntimeError(L('Failed to create user identity: {reportError}', LError(e)))
+    }
+  },
+  'gi.app/identity/signup': async function ({ username, email, password }, publishOptions) {
+    try {
+      const randomAvatar = sbp('gi.utils/avatar/create')
+      const userID = await sbp('gi.app/identity/create', {
+        data: {
+          username,
+          email,
+          password,
+          picture: randomAvatar
+        },
+        publishOptions
+      })
+      return userID
+    } catch (e) {
+      console.error('gi.app/identity/signup failed!', e)
+      await sbp('gi.app/identity/logout') // TODO: should this be here?
+      const message = LError(e)
+      if (e.name === 'GIErrorUIRuntimeError') {
+        // 'gi.app/identity/create' also sets reportError
+        message.reportError = e.message
+      }
+      throw new GIErrorUIRuntimeError(L('Failed to signup: {reportError}', message))
+    }
+  },
+  'gi.app/identity/login': async function ({ username, password: wpassword, identityContractID }: {
+    username: ?string, password: ?Secret<string>, identityContractID: ?string
+  }) {
+    if (username) {
+      identityContractID = await sbp('namespace/lookup', username)
+    }
+
+    if (!identityContractID) {
+      throw new GIErrorUIRuntimeError(L('Incorrect username or password'))
+    }
+
+    const password = wpassword?.valueOf()
+    const transientSecretKeys = []
+    if (password) {
+      try {
+        const salt = await sbp('gi.app/identity/retrieveSalt', username, wpassword)
+        const IEK = await deriveKeyFromPassword(CURVE25519XSALSA20POLY1305, password, salt)
+        transientSecretKeys.push(IEK)
+      } catch (e) {
+        console.error('caught error calling retrieveSalt:', e)
+        throw new GIErrorUIRuntimeError(L('Incorrect username or password'))
+      }
+    }
+
+    try {
+      sbp('appLogs/startCapture', identityContractID)
+      const { state, cheloniaState, encryptionParams } = await loadState(identityContractID, password)
+      let loginCompleteHandler, loginErrorHandler
+
+      try {
+        const loginCompletePromise = new Promise((resolve, reject) => {
+          const loginCompleteHandler = ({ identityContractID: id }) => {
+            sbp('okTurtles.events/off', LOGIN_ERROR, loginErrorHandler)
+            if (id === identityContractID) {
+              // We need to save the state to ensure it's possible to refresh
+              // the page
+              resolve(sbp('state/vuex/save'))
+            } else {
+              reject(new Error('Unexpected identity contract ID'))
+            }
+          }
+          const loginErrorHandler = ({ identityContractID: id, error }) => {
+            sbp('okTurtles.events/off', LOGIN_COMPLETE, loginCompleteHandler)
+            if (id === identityContractID) {
+              reject(error)
+            } else {
+              reject(new Error('Unexpected identity contract ID'))
```

Comment:
The error message should say what the mismatch was. E.g.:

```js
new Error(`Identity contract ID mismatch during login: ${identityContractID} != ${id}`)
```

Reply:
Updated